contracts/collections/sg721-base/src/contract.rs

```diff
@@ -366,40 +380,50 @@ where
-        // make sure the correct contract is being upgraded, and it's being
-        // upgraded from the correct version.
+    pub fn migrate(mut deps: DepsMut, env: Env, _msg: Empty) -> Result<Response, ContractError> {
+        let prev_contract_version = cw2::get_contract_version(deps.storage)?;
+
+        let valid_contract_names = vec![CONTRACT_NAME.to_string()];
+        if !valid_contract_names.contains(&prev_contract_version.contract) {
+            return Err(StdError::generic_err("Invalid contract name for migration").into());
+        }
 
-        if CONTRACT_VERSION < EARLIEST_VERSION {
-            return Err(
-                StdError::generic_err("Cannot upgrade to a previous contract version").into(),
-            );
+        #[allow(clippy::cmp_owned)]
+        if prev_contract_version.version >= CONTRACT_VERSION.to_string() {
+            return Err(StdError::generic_err("Must upgrade contract version").into());
         }
-        if CONTRACT_VERSION > TO_VERSION {
-            return Err(
-                StdError::generic_err("Cannot upgrade to a previous contract version").into(),
-            );
+
+        let mut response = Response::new();
+
+        #[allow(clippy::cmp_owned)]
+        if prev_contract_version.version < "3.0.0".to_string() {
+            response = crate::upgrades::v3_0_0::upgrade(deps.branch(), &env, response)?;
         }
-        // if same version return
-        if CONTRACT_VERSION == TO_VERSION {
-            return Ok(Response::new());
+
+        #[allow(clippy::cmp_owned)]
+        if prev_contract_version.version < "3.1.0".to_string() {
+            response = crate::upgrades::v3_1_0::upgrade(deps.branch(), &env, response)?;
         }
 
-        // update contract version
-        cw2::set_contract_version(deps.storage, CONTRACT_NAME, TO_VERSION)?;
+        cw2::set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+        response = response.add_event(
+            Event::new("migrate")
+                .add_attribute("from_name", prev_contract_version.contract)
+                .add_attribute("from_version", prev_contract_version.version)
+                .add_attribute("to_name", CONTRACT_NAME)
+                .add_attribute("to_version", CONTRACT_VERSION),
+        );
 
-        // perform the upgrade
-        let cw17_res = cw721_base::upgrades::v0_17::migrate::<Extension, Empty, Empty, Empty>(deps)
-            .map_err(|e| ContractError::MigrationError(e.to_string()))?;
-        let mut sgz_res = Response::new();
-        sgz_res.attributes = cw17_res.attributes;
-        Ok(sgz_res)
+        Ok(response)
     }
 }
 
 pub fn share_validate(share: Decimal) -> Result<Decimal, ContractError> {
     if share > Decimal::one() {
-        return Err(ContractError::InvalidRoyalties {});
+        return Err(ContractError::InvalidRoyalties(
+            "Share cannot be greater than 100%".to_string(),
+        ));
     }
```

Comment:
Decimal type allows only positive numbers right ?

Reply:
yes Decimal is Uint under the hood

<img width="1148" alt="Screenshot 2023-07-31 at 11 07 18 AM" src="https://github.com/public-awesome/launchpad/assets/6496257/bf8ebf61-8cb5-4760-809a-157c727206db">